### ISSUE TITLE
Improve interaction with winner-undo

### ISSUE THREAD

**Issue by timor** [ch11ng/exwm#709]
### Motivation
When using `winner-mode`, `winner-undo` checks `last-command` if it should continue skipping back, or consider the chain of undo operations to be complete.  Since key-press events in exwm buffers are not handled by the command loop, the value of `last-command` does not change.  This causes `winner-undo` to continue skipping back even though keys were pressed after the last `winner-undo` operation.

Semantically, I think interacting with an application in an EXWM buffer is similar to editing/navigating commands in regular Emacs buffers.  It would be nice if this behavior could be adjusted accordingly.

### Reproduction
- Have key binding for `winner-undo`

This is the existing behavior with regular buffers, which is expected to work with EXWM buffers:
- Switch between some buffers, end up in buffer A
- perform a `winner-undo`, end up in buffer B
- edit/navigate in buffer B
- perform a `winner-undo`, end up in buffer A

This is the existing behavior with EXWM buffers:
- be in regular buffer D
- switch to an EXWM buffer E (e.g. some terminal)
- switch to an EXWM buffer F
- perform a `winner-undo`, end up in buffer E
- press some keys
- perform a `winner-undo`, end up in buffer D (instead of back in F)

**Comment by ch11ng** [comment]
Clearing the value of `last-command` or setting it to some other command would perhaps do the trick (UPDATE: as I just tried, modifying `last-command` doesn't seem to get `winner-undo` work.  I'm wondering why.), but IDK if there is any side effect.  Obviously the said action cannot follow each buffer switch but certain events, or consecutive `winner-undo` would fail.  There probably won't be a solution for char-mode I think.

**Comment by timor** [comment]
> Clearing the value of `last-command` or setting it to some other command would perhaps do the trick (UPDATE: as I just tried, modifying `last-command` doesn't seem to get `winner-undo` work. I'm wondering why.)

I tried something like that, I think it could have worked if I had found the correct place to modify `last-command`. Whenever I tried it it was too eager, so that switching back along the chain was not working anymore isntead.  It would have to be unset whenever a key press is forwarded to an application, but not e.g. when the `super` key is pressed as part of a non-application-related key binding.  I was unable to find that exact place in the code.

> There probably won't be a solution for char-mode I think.

I use line-mode almost exclusively, so that would not be an issue for me.

**Comment by ch11ng** [comment]
Thanks for the info! I'll try later.

**Comment by ch11ng** [comment]
It turns out setting `last-command` alone is not enough.  Winner mode internally captures window configuration each time `post-command-hook` is run.  So it's necessary to also run that hook.  Please check out ed0f63327dbfa7ef2ebe1dc37bfc38d40d531c29 for the fix.

**Comment by timor** [comment]
Thanks a lot for working on this!

I tried out ed0f633 , and the problem in the description is solved.  However, it seems that the `winner-redo` functionality is not working correctly yet.  IIRC, `winner` relies on no other command being made between successive `winner-undo` calls, so that a `winner-redo` will undo all the `winner-undo` calls.

When I previously tried to fix it (`winner-undo` and `winner-redo` bound to `s-u` and `s-U`, respectively), I think the problem was that hitting the key bindings for `winner-redo` was breaking the `winner-undo` last-command chain.  In fact, already pressing and releasing the `Super` key was issuing one of the fake commands, I think.  I don't know if this is also an issue after ed0f633.


tl;dr
Repro:
1. bind `winner-undo` to `s-u` and `winner-redo` to `s-U`
1. `s-u` repeatedly, then:
   1. If stopping in a non-EXWM buffer, hitting `s-U` will correctly perform `winner-redo`
   1. If stopping in an EXWM buffer, hitting `s-U` will report "user-error: Previous command was not a ‘winner-undo’"

**Comment by ch11ng** [comment]
Could you try the patch below and see if there's anything else not working with winner mode?

```diff
diff --git a/exwm-input.el b/exwm-input.el
index 82ce82a4e..a891710da 100644
--- a/exwm-input.el
+++ b/exwm-input.el
@@ -688,9 +688,10 @@ Current buffer must be an `exwm-mode' buffer."
               (set-transient-map `(keymap (t . ,#'exwm-input--noop)))
               (exwm-input--unread-event event))
           ;; Fool some packages into thinking there is a change in the buffer.
-          (setq last-command #'exwm-input--noop)
-          (run-hooks 'pre-command-hook)
-          (run-hooks 'post-command-hook)))
+          (when event
+            (setq last-command #'exwm-input--noop)
+            (run-hooks 'pre-command-hook)
+            (run-hooks 'post-command-hook))))
       (xcb:+request exwm--connection
           (make-instance 'xcb:AllowEvents
                          :mode mode
```

**Comment by timor** [comment]
I tried it out, and the first quick tests seem to work.  I'll use it for a bit and see whether there are any issues with this, but it looks good so far, thanks a lot!

**Comment by medranocalvo** [comment]
@timor, can this ticket be closed?

**Comment by timor** [comment]
Yes, I think this can be closed.  So far it seems to work well.

**Comment by medranocalvo** [comment]
Reopening, since this is not yet installed in the code (@ch11ng).

**Comment by ch11ng** [comment]
Just pushed.

**Comment by timor** [comment]
There is one thing which hasn't been accounted for, but I don't know whether this can be fixed in this scope:  If I only use the mouse in an X window, it will not be registered as having been "active", only key presses cause the window to move to the front of the winner-undo ring.

Should I open a separate issue for this?

**Comment by ch11ng** [comment]
I was not aware of such scenario.  A fix is coming.

**Comment by ch11ng** [comment]
Please checkout 0368127976bda29d35eed788edfe74644ecd3845 for the fix.

**Comment by medranocalvo** [comment]
@timor, the fix by @ch11ng is included in the 0.25 (deabe9c2d) release.

Please check it out and reopen if needed.

Thank you for the thorough report.
